src/Rules/BinaryDist.hs

```diff
@@ -106,16 +112,174 @@ bindistMakefile = unlines
+  , "# names. For hadrian build this will work as programs have a consistent "
+  , "# naming procefure. This file is tested on Linux(Ubuntu)"
+  , "# TODO : Check implementation in other distributions" 
+  , "\trm -f $2"
+  , "\t$(CREATE_SCRIPT) $2"
+  , "\t@echo \"#!$(SHELL)\" >>  $2"
+  , "\t@echo \"exedir=\\\"$4\\\"\" >> $2"
+  , "\t@echo \"exeprog=\\\"$1\\\"\" >> $2"
+  , "\t@echo \"executablename=\\\"$5\\\"\" >> $2"
+  , "\t@echo \"bindir=\\\"$3\\\"\" >> $2"
+  , "\t@echo \"libdir=\\\"$6\\\"\" >> $2"
+  , "\t@echo \"docdir=\\\"$7\\\"\" >> $2"
+  , "\t@echo \"includedir=\\\"$8\\\"\" >> $2"
+  , "\t@echo \"\" >> $2 "
+  , "\tcat wrappers/$1 >> $2" 
+  , "\t$(EXECUTABLE_FILE) $2 ;"
+  , "endef"
+  , ""
+  , "# QUESTION : should we use shell commands?"
+  , ""
+  , "# Due to the fact that package database is configured relatively"
+  , "# We do not change the relative paths of executables and libraries"
+  , "# But instead use wrapper scripts whenever necessary"
+  , "LIBPARENT = $(shell dirname $(libdir))"
+  , "GHCBINDIR = \"$(LIBPARENT)/bin\""
+  , ""
   , ".PHONY: install"
-  , "install:"
-  , "\tmkdir -p $(prefix)"
-  , "\tcp settings lib/settings"
-  , "\tcp -R bin $(prefix)/"
-  , "\tcp -R lib $(prefix)/"
+  , "install: install_bin install_lib"
+  , ""
+  , "# Check if we need to install docs"
+  , "ifeq \"DOCS\" \"YES\""
+  , "install: install_docs"
+  , "endif"
+  , ""
+  , "# If the relative path of binaries and libraries are altered, we will need to"
+  , "# install additional wrapper scripts at bindir."   
+  , "ifneq \"$(LIBPARENT)/bin\" \"$(bindir)\""
+  , "install: install_wrappers"
+  , "endif"
+  , ""
+  , "# We need to install binaries relative to libraries."
+  , "BINARIES = $(wildcard ./bin/*)" 
+  , "install_bin:"
+  , "\t@echo \"Copying Binaries to $(GHCBINDIR)\""
+  , "\t$(INSTALL_DIR) \"$(GHCBINDIR)\""
+  , "\tfor i in $(BINARIES); do \\"
+  , "\t\tcp -R $$i \"$(GHCBINDIR)\"; \\"
+  , "\tdone"
+  , ""
+  , "LIBRARIES = $(wildcard ./lib/*)"
+  , "install_lib:"
+  , "\t@echo \"Copying libraries to $(libdir)\""
+  , "\t$(INSTALL_DIR) \"$(libdir)\""
+  , "\tfor i in $(LIBRARIES); do \\"
+  , "\t\tcp -R $$i \"$(libdir)/\"; \\"
+  , "\tdone"
+  , ""
+  , "DOCS = $(wildcard ./docs/*)"
+  , "install_docs:"
+  , "\t@echo \"Copying libraries to $(docdir)\""
+  , "\t$(INSTALL_DIR) \"$(docdir)\""
+  , "\tfor i in $(DOCS); do \\"
+  , "\t\tcp -R $$i \"$(docdir)/\"; \\"
+  , "\tdone"
+  , ""
+  , "BINARY_NAMES=$(shell ls ./bin/)"
+  , "install_wrappers:"
+  , "\t@echo \"Installing Wrapper scripts\""
+  , "\t$(INSTALL_DIR) \"$(bindir)\""
+  , "\t$(foreach p, $(BINARY_NAMES),\\"
+  , "\t\t$(call installscript,$p,$(bindir)/$p,$(bindir),$(GHCBINDIR),$(GHCBINDIR)/$p,$(libdir),$(docdir),$(includedir)))"
+  , ""
+  , "# END INSTALL"
+  , "# -----------------------------------------------------------------------------"
   ]
+
+basename :: FilePath -> String
+basename path = last $ splitOn "/" path
```

Comment:
Should this be replaced with `takeFileName`?

Reply:
I mean the standard function:

https://hackage.haskell.org/package/filepath-1.4.2/docs/System-FilePath-Posix.html#v:takeFileName

Reply:
Resolved. Thanks @snowleopard